doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+
+### Permissions management for app passwords
+
+A static set of new ACIs must be added, which will ensure that the owning user
+will have read-write access to all (and only) objects holding his
+app passwords, as well as appropriate access to their attributes - both when
+logged in using the primary password and when logged in using the respective app
+password. This is the only thing to do with respect to ACIs and app passwords,
+as ACIs serve only for work with the LDAP database.
+
+The only general limitation when the user is logged in using an app password
+will be the inability to manage the account (modify the information about the
+user, change his other passwords...). As existing ACIs will not apply to users
+logged in using an app password, even a limitation such as reducing the
+effective permissions of the current user session (possibly according to the app
+policy) using a 389-DS plugin is not needed.
+
+### SSSD support for the aforementioned functionality
+
+Interactions with SSSD using the new functionality must be tested in
+`ipatests/test_integration/test_sssd.py`.
+
+If needed, SSSD will also be extended to support the aforementioned
+functionality. This will be done by a closely related, but separate extension of
+SSSD.
+
+## Implementation
+
+The extension must be backward compatible, that means not breaking any existing
+functionality and passing all existing tests.
+
+As for changes to the data storage:
+* As described above, a new LDAP subtree will be created for app passwords, with
+  objects holding them being grouped by app, and their attributes being, besides
+  an app password itself for LDAP, a uid, a descriptive display name, an app
+  name, a UUID, and a backlink to the owning user.
+* New ACIs will be created so that users will have needed access to objects
+  holding their own app passwords.
+
+A new objectclass will be created for the `apppw` object and added into
+the LDAP schema. This objectclass will be defined in
+`install/share/60basev3.ldif`:
+```
+objectClasses:
+    (2.16.840.1.113730.3.8.12.XX
+    NAME 'ipaAppPw'
+    DESC 'Object containing an app password and related data'
+    SUP top
+    MUST ( uid $ displayName $ appName $ entryUUID $ managedBy )
+    X-ORIGIN 'IPA v4.9')
+```
+
+The `userPassword` attribute will be added by including the
+`simpleSecurityObject` into the `apppw` object.
+
+The LDIF code to add the `cn=apps,cn=accounts,$SUFFIX` entry, as well as new
+ACIs for access to the `apppw` object and its attributes, during
+* new installations will be in `install/share/apppw.ldif`,
+* upgrades will be in `install/updates/19-apppw.update` (entry) and
+  `install/updates/20-aci.update` (ACIs).
+
+There will be the following ACIs on the `cn=apps,cn=accounts,$SUFFIX` entry:
+```
+aci: (target = "cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (version 3.0; acl "Allow users to add entries grouping app passwords for a
+     particular app";
+     allow (add) 
+     userdn = "ldap:///all;)
```

Comment:
I wonder if some admins would push back at users being able to create entries like this.

Reply:
Ignore, that's the design. User-defined apps. Admins will be able to intervene if someone goes absolutely nuts.

Reply:
This is my tentative proposal, as such an entry must exist for the user to be able to create the object holding the new app password. And as discussed [above](https://github.com/freeipa/freeipa/pull/4061#issuecomment-640033139) (and as I see below, you noticed it afterwards), it is even possible that there will instead be entries grouping the app password objects by users, not by apps. This would solve this particular problem, but it is still in open discussion. If we proceed to group these objects by apps, then I don't know how to achieve that in other, more elegant and secure way than this. Will we require that these entries will be created by admins? That would be impractical. It starts to seem that grouping these objects by users may indeed be a better way to go - as noted [above](https://github.com/freeipa/freeipa/pull/4061#issuecomment-640033139) (in response to point 9), it would also have another benefit.

Reply:
All right, I did not see your second post when writing my reaction. It remains to decide on the idea to group app passwords by users instead of by apps.